src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs

```diff
@@ -0,0 +1,503 @@
+            Debug.Assert(!Monitor.IsEntered(_gate));
+
+            // Track the time the we start waiting.
+            long startTime = GetTimestamp();
+
+            // Polling loop
+            while (true)
+            {
+                bool createdTask = false;
+                CancellationTokenSource cts = null;
+                Task waitTask;
+
+                // We're in a polling loop... determine how much time remains
+                int remainingTimeout = millisecondsTimeout == Timeout.Infinite ?
+                    Timeout.Infinite :
+                    (int)Math.Max(millisecondsTimeout - (GetTimestamp() - startTime), 0);
+
+                lock (_gate)
+                {
+                    // If we already know that the process exited, we're done.
+                    if (_exited)
+                    {
+                        return true;
+                    }
+
+                    // If a timeout of 0 was supplied, then we simply need to poll
+                    // to see if the process has already exited.
+                    if (remainingTimeout == 0)
+                    {
+                        // If there's currently a wait-in-progress, then we know the other process
+                        // hasn't exited (barring races and the polling interval).
+                        if (_waitInProgress != null)
+                        {
+                            return false;
+                        }
+
+                        // No one else is checking for the process' exit... so check.
+                        // We're currently holding the _gate lock, so we don't want to
+                        // allow CheckForExit to block indefinitely.
+                        CheckForExit();
+                        return _exited;
+                    }
+
+                    // The process has not yet exited (or at least we don't know it yet)
+                    // so we need to wait for it to exit, outside of the lock.
+                    // If there's already a wait in progress, we'll do so later
+                    // by waiting on that existing task.  Otherwise, we'll spin up
+                    // such a task.
+                    if (_waitInProgress != null)
+                    {
+                        waitTask = _waitInProgress;
+                    }
+                    else
+                    {
+                        createdTask = true;
+                        CancellationToken token = remainingTimeout == Timeout.Infinite ?
+                            CancellationToken.None :
+                            (cts = new CancellationTokenSource(remainingTimeout)).Token;
+                        waitTask = WaitForExitAsync(token);
+
+                        // PERF NOTE:
+                        // At the moment, we never call CheckForExit with blockingAllowed:true (which in turn allows
+                        // waitpid to block until the child has completed) because we currently allows call it while
```

Comment:
I think "because we currently allow call it" should be "because we currently call it"